core-primitives/stf-executor/src/executor.rs

```diff
@@ -0,0 +1,441 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+// #[cfg(all(not(feature = "std"), feature = "sgx"))]
+// use crate::sgx_reexport_prelude::*;
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use std::untrusted::time::SystemTimeEx;
+
+use crate::{
+	error::{Error, Result},
+	traits::{
```

Comment:
This is the design question of how many methods should a trait have. In other languages, recommendation for interfaces are in the range of 2-5 or 2-7. With traits, where we have trait mix-in (with `+`) that makes trait aggregation very easy, we can achieve very fine granularity by having one method per trait.
In our case here, most consumers don't ever need more than 2 of these methods. So grouping them together in a trait, will cause unnecessary exposure of methods. That was my reasoning behind this design. However, I think this can be open for debate, I'd be happy to hear your arguments 😄

Reply:
We'd might want to do __ what with these traits? I think you lost the verb there in the sentence 😅

Reply:
haha, thought faster than I typed. 😄 I meant 'remove'. However, not in terms of combining traits, but with a better core design. :P

Reply:
Ah okay, now it makes sense 😝. Yes I totally agree, I think these stf executor traits can be reduced in number, because they are somewhat redundant and can be combined into more generalized trait methods. However, I believe refactoring that requires a bit of effort, it's not that straight forward (and will require writing some tests in advance, to make sure we don't break stuff in the process).

Reply:
My reasoning is just to keep it simple. The stf is made to be adaptable by clients such that they can easily integrate their stuff into the sgx-runtime. As long as we keep in mind the users who want to remove unshielding & shielding, or whatever, I'm fine with it. And it's not like I have a better solution here myself ;)

Reply:
We can certainly make it more simple. What I have here in this PR is just extracting code as-is into a crate. It is by no means the end-result as we would want to have it. There are still many improvements that can be done on this part, this was just a first step.